Wallet, GUI: Warn when sending to already-used Bitcoin addresses

luke-jr opened the pull request:
- An in-memory bloom filter is used to detect potential address reuse, avoiding wasting unnecessary memory with large wallets.
- Entering a used address in the GUI Send tab makes the field turn yellow.
- Sending to a used address from the GUI prompts with detailed information about prior usage, as well as a note about best practices to avoid address reuse.
- (I also fixed `GUIUtil::dateTimeStr` to not overflow with 64-bit timestamps.)

luke-jr committed:
Wallet: Add memory-only bloom filter for addresses used in wallet transactions

luke-jr committed:
Wallet: Add efficient [negative] check that an address is not known to be used

luke-jr committed:
GUI: QValidatedLineEdit: Add support for a warning-but-valid state

luke-jr committed:
interfaces/wallet: Add checkAddressForUsage and findAddressUsage

luke-jr committed:
GUI: WalletModel: Wrap checkAddressForUsage and findAddressUsage

luke-jr committed:
GUI: Implement BitcoinAddressUnusedInWalletValidator

luke-jr committed:
GUI: Use warning indicator for send coins entries with reused addresses

luke-jr committed:
GUI: Use 64-bit safe QDateTime::fromMSecsSinceEpoch for dateTimeStr

Partially reverts 0030c1bd6cdee44d9f59af9a023cd39cbd335bd5, which was presumably done for compatibility with Qt<4.7.

luke-jr committed:
GUI: Add GUIUtil::dateStr

luke-jr committed:
GUI: Add a warning prompt when sending to an already-used address

gmaxwell commented:
Meta-concept-ack!  we should absolutely do something like this (I haven't looked at the specifics of what this does yet)

luke-jr commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -243,6 +245,72 @@ void SendCoinsDialog::on_sendButton_clicked()
+            const QString label_and_address = rcp.label.isEmpty() ? rcp.address : (rcp.label + " (" + rcp.address + ")");
+            reuse_question.append("<br />");
+            if (rcp_prior_usage_info.num_txs == 1) {
+                //: %1 is an amount (eg, "1 BTC"); %2 is a Bitcoin address and its label; %3 is a date (eg, "2019-05-08")
```
Note: `//:` is how Qt lets us add notes for translators. (I'm not sure if it survives to Transifex?)

hebasto replied:
> Note: `//:` is how Qt lets us add notes for translators. (I'm not sure if it survives to Transifex?)

Smth like #21465 is required for that.

Rockstarrecords11 commented:
Give it up

Rockstarrecords11 commented on src/wallet/wallet.h:
```diff
@@ -851,6 +857,8 @@ class CWallet final : public CCryptoKeyStore, public CValidationInterface
+
     /*
      * Rescan abort properties
      */
```
Give to me

meshcollider commented:
Concept ACK

This addresses #3266 I assume

luke-jr commented:
It doesn't take into consideration all the ideas/advice (even my own!) on #3266, but yes, it implements the general idea I think.

jonasschnelli commented:
Concept ACK

hebasto commented:
Concept ACK

sipa commented on src/wallet/wallet.cpp:
```diff
@@ -723,6 +723,23 @@ void CWallet::AddToSpends(const uint256& wtxid)
+
+void CWallet::BuildAddressBloomFilter()
+{
+    m_address_bloom_filter_elems = std::max<size_t>(100, mapWallet.size() * 2);
```
Is this code assuming that there are two outputs per transaction? Bloom filter performance is very much dependent on this estimate being correct (if you take a filter designed for N elements with fprate=0.000001, and put 2N elements in it, the resulting fprate is only 0.0032).

luke-jr replied:
The intent was to not need rebuilding until the wallet doubled in size, but I guess it fails to consider transactions having multiple outputs.

Would another factor (or variable) be better here?

luke-jr replied:
Made it 6x for now.

Sjors replied:
I'm still seeing `* 2`. Move the constants `ADDR_BLOOM_FILTER_TX_TO_ELEMENTS_FACTOR` and `ADDR_BLOOM_FILTER_FP_RATE` defined in 39f49f653541ce172a91a8d6996f44831f9a8df2 to  a13c30296f0507b1c6cdd70e42e693fe38d3c9df?

promag commented:
Concept ACK, didn't see the code but maybe you could split RPC changes to other PR?

Sjors commented:
Concept ACK, but agree with @promag on splitting `getaddressinfo` into a seperate PR, so we can review that and the bloom filter stuff first.

When I enter a duplicate address the field becomes yellow as expected, but when I also enter an absurdly high amount, it no longer shows the "insufficient balance error", but instead ignores the amount I entered and falls back to the previous amount. (I tried this on a rebased branch, so maybe I broke it myself)

luke-jr committed:
GUI: SendConfirmationDialog: Enable customisation of dialog

- Dialog icon can be changed
- Both buttons can be replaced with other standard buttons
- "Yes" button can be renamed

luke-jr commented:
Rebased, fixed issues, and moved RPC change to a new `rpc_gai_txids` branch that can be PR'd after this.

luke-jr committed:
GUI/Send: Use detailedText for address reuse error

instagibbs commented:
concept ACK

Sjors commented on src/wallet/wallet.cpp:
```diff
@@ -841,6 +841,23 @@ void CWallet::AddToSpends(const uint256& wtxid)
 }
 
+void CWallet::AddToAddressBloomFilter(const CWalletTx& wtx)
+{
```
Assert that `m_address_bloom_filter` has been built?

Sjors commented on src/wallet/wallet.h:
```diff
@@ -1019,6 +1019,8 @@ class CWallet final : public FillableSigningProvider, private interfaces::Chain:
     void UnlockAllCoins() EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
     void ListLockedCoins(std::vector<COutPoint>& vOutpts) const EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
 
+    bool FindScriptPubKeyUsed(const std::set<CScript>& keys, const boost::variant<boost::blank, std::function<void(const CWalletTx&)>, std::function<void(const CWalletTx&, uint32_t)>>& callback = boost::blank()) const EXCLUSIVE_LOCKS_REQUIRED(cs_wallet);
```
Can you document the params here? Also maybe explicitly state that this function corrects for any false positives found with the bloom filter.

I think using `std::vector<CWalletTx&>& transactions` instead of a callback is more readable (and avoids Boost). Or do you need this to be asynchronous?

Sjors commented on src/wallet/wallet.cpp:
```diff
@@ -850,14 +853,49 @@ void CWallet::AddToAddressBloomFilter(const CWalletTx& wtx)
+    if (!found_any) {
+        return false;
+    }
+
```
Suggested comment: `// Bloom filters can return false positives, so iterate through all wallet addresses`

promag replied:
Could early return if no callback?

Sjors commented on src/interfaces/wallet.cpp:
```diff
@@ -30,6 +31,16 @@
 namespace interfaces {
 namespace {
 
+std::set<CScript> AddressesToKeys(std::vector<std::string> addresses)
```
`AddressesToScriptPubKeys`?

Sjors commented on src/qt/walletmodel.h:
```diff
@@ -158,6 +160,8 @@ class WalletModel : public QObject
 
     // Check address for validity
     bool validateAddress(const QString &address);
+    bool checkAddressForUsage(const std::vector<std::string>& addresses) const;
```
`checkAddressesForUsage`?

Sjors commented on src/qt/sendcoinsdialog.cpp:
```diff
@@ -274,6 +276,71 @@ void SendCoinsDialog::on_sendButton_clicked()
+            QStringList addresses;
+            for (const auto& recipient : recipients) {
+#ifdef ENABLE_BIP70
+                if (recipient.paymentRequest.IsInitialized()) continue;
```
Maybe add a comment here. Why not put `ifndef ENABLE_BIP70` before the entire `{` block? Can users manually add destinations to a BIP70 payment?

fanquake replied:
> Can users manually add destinations to a BIP70 payment?

BIP70 is no longer supported. Looks like this needs to be rebased on master and any `ENABLE_BIP70` additions needs to be removed.

Sjors reviewed (commented):
391c5d9a97211b46bc6b4d701e98bf10aae197fa looks good, except the triplet of choices is confusing:

<img width="454" alt="Schermafbeelding 2019-11-15 om 18 18 22" src="https://user-images.githubusercontent.com/10217/68962213-547a8c00-07d4-11ea-9a1b-ac19cbec6cad.png">

* try renaming `OK` to `Cancel` and making that the default action
* `Override` should be `Send` or `Send anyway` (I initially thought Override meant overriding the address)
* Consider dropping `Show Details` and instead put `Sent ... BTC to this address across 2 transactions from 15-11-2019 through 15-11-2019` as a text under the address. Being able to put an error message right under the address is useful for a #16807 GUI followup too.

Given 391c5d9a97211b46bc6b4d701e98bf10aae197fa, I think #17463 should be merged first.

promag reviewed (commented):
Would it be too bad to index all wallet's scriptPubKey?

luke-jr commented:
Planning to redo this as an addressbook `bool` once #18192 gets merged...

hebasto commented:
> Planning to redo this as an addressbook `bool` once #18192 gets merged...

#18192 and #18546 have been merged already :D

rebroad commented:
@luke-jr re-open?

hebasto commented:
Up for grabs?

luke-jr commented:
It's maintained. PR is pending on https://github.com/bitcoin-core/gui/pull/404

luke-jr commented:
Now blocking on PR #22693 for wallet changes
